flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/logbundler/LogArchiver.java

```diff
@@ -0,0 +1,146 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.rest.handler.logbundler;
+
+import org.apache.flink.annotation.VisibleForTesting;
+
+import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
+import org.apache.commons.compress.archivers.tar.TarArchiveOutputStream;
+import org.apache.commons.compress.compressors.gzip.GzipCompressorOutputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.StandardOpenOption;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static org.apache.flink.util.Preconditions.checkState;
+
+/** Simple util for creating a log archive. */
+public class LogArchiver {
+    private static final Logger log = LoggerFactory.getLogger(LogArchiver.class);
+    private static final Pattern namePattern = Pattern.compile("\\.([0-9]+)$");
+
+    private final TarArchiveOutputStream archiveOutputStream;
+    private final File archive;
+    private boolean isClosed;
+    private final Set<String> entryNames = new HashSet<>();
+
+    public LogArchiver(File file) throws IOException {
+        this.archive = file;
+
+        OutputStream fo =
+                Files.newOutputStream(
+                        file.toPath(),
+                        StandardOpenOption.CREATE,
+                        StandardOpenOption.WRITE,
+                        StandardOpenOption.TRUNCATE_EXISTING);
+        OutputStream gzo = new GzipCompressorOutputStream(fo);
+        this.archiveOutputStream = new TarArchiveOutputStream(gzo);
+        isClosed = false;
+        entryNames.clear();
+    }
+
+    public File getArchive() {
+        checkState(isClosed, "Can not download archive before it has been closed.");
+        return archive;
+    }
+
+    public synchronized void addArchiveEntry(
+            String entryName, InputStream input, long size, long modTime) throws IOException {
+        checkState(!isClosed, "Can not add archive entry to closed archiver");
+        String internalEntryName = entryName;
+        while (entryNames.contains(internalEntryName)) {
+            internalEntryName = getNextEntryName(internalEntryName);
+        }
+        log.debug("Adding archive entry name:{} size:{}", internalEntryName, size);
+        entryNames.add(internalEntryName);
+        TarArchiveEntry entry = new TarArchiveEntry(internalEntryName);
+        entry.setSize(size);
+        entry.setMode(TarArchiveEntry.DEFAULT_FILE_MODE);
+        entry.setModTime(modTime);
+        entry.setUserName("");
+        try {
+            archiveOutputStream.putArchiveEntry(entry);
+            // copy with limit: the input log file could have additional data, which would confuse
+            // the archiver
```

Comment:
This comment was indeed sloppy. I added a more extensive one to the method:
```java
    /**
     * Special copy method which defines a "reading window" on the input stream. The
     * "TarArchiveOutputStream" expects an TarArchiveEntry with a defined size. If the log file is
     * growing while the entry is added to the output stream, it will throw an exception. This
     * method ensures that only "limit" bytes will be written from the "input" to the "output"
     * stream.
     */
```

Reply:
The only case where the file can grow are the JM logs, correct?

Reply:
Exactly.

Reply:
OK then the limit makes sense.
But it does make me wonder how well this approach works when the log files roll over; the contents may suddenly change, as does the file size.

Reply:
(How well does the log transmission from the TM handle rollovers?)